File: library/src/main/java/name/zeno/android/util/MeasureUtils.java
package name.zeno.android.util;

import android.graphics.Point;
import android.view.View;

/**
 * Create Date: 16/6/19
 *
 * @author 陈治谋 (513500085@qq.com)
 */
public class MeasureUtils
{
  public static Point as4Rate3(int widthMeasureSpec, int heightMeasureSpec)
  {
    int wm = View.MeasureSpec.getMode(widthMeasureSpec);
    int hm = View.MeasureSpec.getMode(heightMeasureSpec);
    int w, h;
    switch (wm) {
      case View.MeasureSpec.AT_MOST:
      case View.MeasureSpec.EXACTLY:
        w = View.MeasureSpec.getSize(widthMeasureSpec);
        break;
      case View.MeasureSpec.UNSPECIFIED:
      default:
        w = ZDimen.getWindowPixelsSize().x;
        break;
    }

    switch (hm) {
      case View.MeasureSpec.AT_MOST:
      case View.MeasureSpec.EXACTLY:
        h = View.MeasureSpec.getSize(heightMeasureSpec);
        break;
      case View.MeasureSpec.UNSPECIFIED:
      default:
        h = ZDimen.getWindowPixelsSize().y;
        break;
    }
    if (w * 3 / 4 > h) {
      w = h * 4 / 3;
    }
    if (h > w * 3 / 4) {
      h = w * 3 / 4;
    }

    return new Point(w, h);
  }

  public static Point full(int widthMeasureSpec, int heightMeasureSpec)
  {
    int wm = View.MeasureSpec.getMode(widthMeasureSpec);
    int hm = View.MeasureSpec.getMode(heightMeasureSpec);
    int w, h;
    switch (wm) {
      case View.MeasureSpec.AT_MOST:
      case View.MeasureSpec.EXACTLY:
        w = View.MeasureSpec.getSize(widthMeasureSpec);
        break;
      case View.MeasureSpec.UNSPECIFIED:
      default:
        w = ZDimen.getWindowPixelsSize().x;
        break;
    }

    switch (hm) {
      case View.MeasureSpec.AT_MOST:
      case View.MeasureSpec.EXACTLY:
        h = View.MeasureSpec.getSize(heightMeasureSpec);
        break;
      case View.MeasureSpec.UNSPECIFIED:
      default:
        h = ZDimen.getWindowPixelsSize().y;
        break;
    }

    return new Point(w, h);

  }
}

File: library/src/main/java/name/zeno/android/util/MeasureUtils.kt
package name.zeno.android.util

import android.graphics.Point
import android.view.View

/**
 * Create Date: 16/6/19
 *
 * @author 陈治谋 (513500085@qq.com)
 */
object MeasureUtils {
  fun as4Rate3(widthMeasureSpec: Int, heightMeasureSpec: Int): Point {
    val wm = View.MeasureSpec.getMode(widthMeasureSpec)
    val hm = View.MeasureSpec.getMode(heightMeasureSpec)
    var w: Int
    var h: Int
    w = when (wm) {
      View.MeasureSpec.AT_MOST, View.MeasureSpec.EXACTLY -> View.MeasureSpec.getSize(widthMeasureSpec)
      View.MeasureSpec.UNSPECIFIED -> ZDimen.windowPixelsSize.x
      else -> ZDimen.windowPixelsSize.x
    }

    h = when (hm) {
      View.MeasureSpec.AT_MOST, View.MeasureSpec.EXACTLY -> View.MeasureSpec.getSize(heightMeasureSpec)
      View.MeasureSpec.UNSPECIFIED -> ZDimen.windowPixelsSize.y
      else -> ZDimen.windowPixelsSize.y
    }
    if (w * 3 / 4 > h) {
      w = h * 4 / 3
    }
    if (h > w * 3 / 4) {
      h = w * 3 / 4
    }

    return Point(w, h)
  }

  fun full(widthMeasureSpec: Int, heightMeasureSpec: Int): Point {
    val wm = View.MeasureSpec.getMode(widthMeasureSpec)
    val hm = View.MeasureSpec.getMode(heightMeasureSpec)
    val w = when (wm) {
      View.MeasureSpec.AT_MOST, View.MeasureSpec.EXACTLY -> View.MeasureSpec.getSize(widthMeasureSpec)
      View.MeasureSpec.UNSPECIFIED -> ZDimen.windowPixelsSize.x
      else -> ZDimen.windowPixelsSize.x
    }

    val h = when (hm) {
      View.MeasureSpec.AT_MOST, View.MeasureSpec.EXACTLY -> View.MeasureSpec.getSize(heightMeasureSpec)
      View.MeasureSpec.UNSPECIFIED -> ZDimen.windowPixelsSize.y
      else -> ZDimen.windowPixelsSize.y
    }

    return Point(w, h)

  }
}
